### DISCUSSION TITLE
Understanding how caching helps in auto-refreshing dashboards

### DISCUSSION THREAD

**Discussion by NikosSpanos** [apache/superset#20803] [Q&A / Help]
I have created a couple of Superset dashboards in a production environment, and I have read that it's recommended to use the Redis cache in production environments. A similar StackOverflow question [here](https://stackoverflow.com/questions/72143877/my-superset-cache-is-set-to-redis-but-i-still-get-cache-warning-in-logs).

**Firstly**, I would like to understand what am I going to achieve by adding the following code in ```superset_config.py```
```
FILTER_STATE_CACHE_CONFIG = {
    'CACHE_TYPE': 'RedisCache',
    'CACHE_DEFAULT_TIMEOUT': 86400,
    'CACHE_KEY_PREFIX': 'superset_filter_',
    'CACHE_REDIS_URL': 'redis://localhost:6379/2'
}
```

**Secondly**, I would like to know how can I *auto-refresh* a superset dashboard permanently and not only *for the current session* option which is available. Is there any way?

Thanks for your help in advance.

-- UPDATE 26.07.2022 (after more research)

Reference links: [official doc][3], [issue 390][4]

I have added the following dictionaries in my ```superset_config.py``` file:

```python
CACHE_CONFIG: CacheConfig = {
    'CACHE_TYPE': 'RedisCache',
    'CACHE_DEFAULT_TIMEOUT': int(timedelta(days=1).total_seconds()),
    'CACHE_KEY_PREFIX': 'superset_cache_',
    'CACHE_REDIS_URL': 'redis://redis:6379/2'
}

# Cache for datasource metadata and query results
DATA_CACHE_CONFIG: CacheConfig = {
    'CACHE_TYPE': 'RedisCache',
    'CACHE_DEFAULT_TIMEOUT': int(timedelta(days=1).total_seconds()),
    'CACHE_KEY_PREFIX': 'superset_data_',
    'CACHE_REDIS_URL': 'redis://redis:6379/3'
}

FILTER_STATE_CACHE_CONFIG: CacheConfig = {
    'CACHE_TYPE': 'RedisCache',
    'CACHE_DEFAULT_TIMEOUT': int(timedelta(days=1).total_seconds()),
    'CACHE_KEY_PREFIX': 'superset_filter_',
    'CACHE_REDIS_URL': 'redis://redis:6379/4'
}

EXPLORE_FORM_DATA_CACHE_CONFIG: CacheConfig = {
    'CACHE_TYPE': 'RedisCache',
    'CACHE_DEFAULT_TIMEOUT': int(timedelta(days=1).total_seconds()),
    'CACHE_KEY_PREFIX': 'superset_explore_',
    'CACHE_REDIS_URL': 'redis://redis:6379/5'
}
```

The superset app started successfully and when I do a dashboard refresh I can see the query running in redis-cli. My concern is that every time I apply filters on the dashboards the data is re-cached. Shouldn't the caching be applied once for every filter in the datasource, so when I apply filters superset won't have to hit the DB to fetch new records.

**Comment by nytai** [comment]
no, caching just prevents issuing the same query again. Computing every possible filter combination up front would be incredibly expensive and defeat the purpose of caching. Caching does not mean eagerly prefetching results.

**Reply by NikosSpanos** [reply]
Thanks for the reply @nytai ..So my guess is that I let the users to make every possible filter combination and after every combination is cached, superset won't run the cached queries again. That makes sense. Thanks fot the clarifications. Do you happen to know any best practices regarding caching in Superset other than the ones I have implemented already?

Appreciate the help. Best regards.

**Comment by vnayagam** [comment]
can caching be enabled even without opening the dashboard, means I need all the data to be loaded before I login,

but now it loads for first time and the 2nd time it takes from cache

**Reply by Aalbedon** [reply]
The Caching page (https://superset.apache.org/docs/installation/cache) used to have a section about using celery beat to warm up the cache. Not sure where it has gone.
You can warm up dashboards based on various strategies based on a schedule.
That way when users open the dashboard first thing in the morning they get a cached version served up.

Be aware that if you use Jinja templated datasets in your dashboard that this may not work as expected.

**Reply by vnayagam** [reply]
Thanks, we have this caching already implemented, what we are looking for is cache without opening the session.

Is there a way to cache without opening the dashboards, so that it is not loaded even for the 1st time ?

**Reply by Aalbedon** [reply]
That's what I'm referring to with celery beat.
It ensures that Superset warms up the cache in the background at a certain time.
Also, the celery beat warm up moved to the Async queries section (https://superset.apache.org/docs/installation/async-queries-celery)

**Reply by vnayagam** [reply]
Thank you

**Reply by Aalbedon** [reply]
I should note that the reason the dashboard cache warm up was removed from the Cache documentation page is because it isn't production-ready (see https://github.com/apache/superset/pull/20269).

**Reply by vnayagam** [reply]
I cannot achive the celery implementation, I see some code commented in superset source code, can you please share a code snippet to achieve this bacjground caching process.

**Comment by vnayagam** [comment]
Can any one help me with superset caching like warm up cache so that when user log in for first time it should fetch from cache
